core/src/main/java/io/questdb/cairo/GeoHashes.java

```diff
@@ -0,0 +1,267 @@
+        // element at index 11 is -1
+        return base32Indexes[c > 47 && c < 123 ? c - 48 : 11];
+    }
+
+    public static long fromBitString(CharSequence bits, int start) throws NumericException {
+        return fromBitString(bits, start, Math.min(bits.length(), ColumnType.GEO_HASH_MAX_BITS_LENGTH + start));
+    }
+
+    public static long fromBitStringNl(CharSequence bits, int start) throws NumericException {
+        int len = bits.length();
+        if (len - start <= 0) {
+            return NULL;
+        }
+        return fromBitString(bits, start, Math.min(bits.length(), ColumnType.GEO_HASH_MAX_BITS_LENGTH + start));
+    }
+
+    public static long fromCoordinatesDeg(double lat, double lon, int bits) throws NumericException {
+        if (lat < -90.0 || lat > 90.0) {
+            throw NumericException.INSTANCE;
+        }
+        if (lon < -180.0 || lon > 180.0) {
+            throw NumericException.INSTANCE;
+        }
+        if (bits < 0 || bits > ColumnType.GEO_HASH_MAX_BITS_LENGTH) {
+            throw NumericException.INSTANCE;
+        }
+        return fromCoordinatesDegUnsafe(lat, lon, bits);
+    }
+
+    public static long fromCoordinatesDegUnsafe(double lat, double lon, int bits) {
+        long latq = (long)Math.scalb((lat + 90.0) / 180.0, 32);
+        long lngq = (long)Math.scalb((lon + 180.0) / 360.0, 32);
+        return Numbers.interleaveBits(latq, lngq) >>> (64 - bits);
+    }
+
+    public static long fromString(CharSequence hash) throws NumericException {
+        return fromString(hash, 0, hash.length());
+    }
+
+    public static long fromString(CharSequence hash, int start, int end) throws NumericException {
+        // no bounds/length/nullity checks
+        long geohash = 0;
+        for (int i = start; i < end; ++i) {
+            geohash = appendChar(geohash, hash.charAt(i));
+        }
+        return geohash;
+    }
+
+    public static long fromStringNl(CharSequence geohash, int start, int length) throws NumericException {
+        if (length <= 0 || geohash == null || geohash.length() == 0) {
+            return GeoHashes.NULL;
+        }
+        return fromString(geohash, start, start + Math.min(length, MAX_STRING_LENGTH));
+    }
+
+    public static void addNormalizedGeoPrefix(long hash, int prefixType, int columnType, final LongList prefixes) throws NumericException {
+        final int bits = ColumnType.getGeoHashBits(prefixType);
+        final int columnSize = ColumnType.sizeOf(columnType);
+        final int columnBits = ColumnType.getGeoHashBits(columnType);
+
+        if (hash == NULL || bits > columnBits) {
+            throw NumericException.INSTANCE;
+        }
+
+        final int shift = columnBits - bits;
+        long norm = hash << shift;
+        long mask = GeoHashes.bitmask(bits, shift);
+        mask |= 1L << (columnSize * 8 - 1); // set the most significant bit to ignore null from prefix matching
+
+        prefixes.add(norm);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;prefixes&#39; (index: 3 | type: LongList). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;columnType&#39;, &#39;hash&#39;. The same parameter type is validated here for example: core/src/main/java/io/questdb/cairo/AbstractIntervalDataFrameCursor.java:265. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).